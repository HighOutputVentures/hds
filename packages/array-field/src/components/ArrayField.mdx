import { Canvas, Meta, Story, ArgsTable } from '@storybook/addon-docs';
import ArrayField from './ArrayField';

<Meta title="Component/Array Field" />

# HDS Array Fields

To install package, run command:

```bash
npm i @highoutput/hds-array-field
```

<Canvas>
  <Story id="components-array-field--default"></Story>
</Canvas>

<ArgsTable />
