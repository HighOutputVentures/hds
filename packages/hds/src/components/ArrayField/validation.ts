import { array, object, SchemaOf, string, InferType } from 'yup';

export interface ArrayInputFieldProps {
  value: string;
}

export const ArrayFieldSchema: SchemaOf<ArrayInputFieldProps> = object()
  .shape({
    input: array(
      object().shape({
        value: string().required('input is required'),
      })
    ),
  })
  .required();

export type ArrayFieldTypeValues = InferType<typeof ArrayFieldSchema>;
