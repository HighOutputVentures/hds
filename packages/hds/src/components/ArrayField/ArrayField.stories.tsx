import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import { ThemeProvider } from '../..';
import ArrayField from './ArrayField';

export default {
  title: 'Components/Form/Array Field',
  component: ArrayField,
} as ComponentMeta<typeof ArrayField>;

const Template: ComponentStory<typeof ArrayField> = args => (
  <ThemeProvider>
    <ArrayField {...args} />
  </ThemeProvider>
);

export const Default = Template.bind({});

Default.args = {
  ...Default.args,
  placeholder: 'Enter an Input',
  defaultValues: {
    input: [
      {
        value: '',
      },
    ],
  },
};
