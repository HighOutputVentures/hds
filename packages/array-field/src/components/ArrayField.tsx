import { AddIcon, DeleteIcon } from '@chakra-ui/icons';
import { Button, Flex, Icon, Text, VStack } from '@chakra-ui/react';
import { TextField } from '@highoutput/hds-forms';
import { yupResolver } from '@hookform/resolvers/yup';
import React, { ReactNode } from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { ArrayFieldSchema, ArrayFieldTypeValues } from './validation';

export interface ArrayFieldProps {
  buttonRemoveChildren?: ReactNode;
  buttonAddChildren?: ReactNode;
  defaultValues: {
    input: ArrayFieldTypeValues[];
  };
  onChange: (data: Record<string, any>) => void;
  onRemove: (data: Record<string, any>) => void;
  onAppend: (data: Record<string, any>) => void;
  onBlur: (data: Record<string, any>) => void;
  maxValue?: number;
  label?: string;
  placeholder?: string;
  isRequired?: boolean;
  __btnTestId?: string;
  __inputTestId?: string;
}

const ArrayField = (props: ArrayFieldProps) => {
  const {
    placeholder,
    defaultValues,
    onChange,
    onRemove,
    onBlur,
    onAppend,
    label,
    isRequired,
    __inputTestId,
    __btnTestId,
    buttonAddChildren,
    buttonRemoveChildren,
    maxValue,
  } = props;
  const { control, register, handleSubmit, formState } = useForm<{
    input: ArrayFieldTypeValues[];
  }>({
    defaultValues,
    resolver: isRequired ? yupResolver(ArrayFieldSchema) : undefined,
  });

  const { fields, append, remove } = useFieldArray({
    control,
    name: 'input',
  });

  const handleAddField = handleSubmit(async (data) => {
    if (maxValue && fields.length <= maxValue - 1) {
      append({
        value: '',
      });
      onAppend(data);
    }
  });

  const handleOnBlurInput = handleSubmit(async (data) => {
    onBlur(data);
  });

  const handleRemove = handleSubmit(async (data) => {
    onRemove(data);
  });

  const handleOnChangeInput = handleSubmit(async (data) => {
    onChange(data);
  });

  return (
    <Flex flexDir="column" gap={2}>
      {label && <Text>{label}</Text>}
      {fields.map((item, idx) => (
        <Flex
          gap={2}
          key={item.id}
          onBlur={handleOnBlurInput}
          onChange={handleOnChangeInput}
        >
          <TextField
            id={'input'}
            placeholder={placeholder ?? ''}
            {...register(`input.${idx}.value`)}
            __fieldTestId={__inputTestId ?? 'hds.array-field.input'}
            error={
              isRequired ? formState.errors?.input?.[idx]?.value?.message : ''
            }
          />
          <VStack>
            <Button
              variant="outline"
              onClick={() => {
                remove(idx);
                handleRemove();
              }}
              data-testid={__btnTestId ?? 'hds.array-field.remove.btn'}
              disabled={fields.length <= 1}
              width="32px"
              height="40px"
            >
              {buttonRemoveChildren ? (
                buttonRemoveChildren
              ) : (
                <Icon as={DeleteIcon} />
              )}
            </Button>
            {idx === fields.length - 1 && (
              <Button
                variant="outline"
                width="32px"
                height="40px"
                data-testid={__btnTestId ?? 'hds.array-field.add.btn'}
                disabled={maxValue === fields.length}
                onClick={handleAddField}
              >
                {buttonAddChildren ? buttonAddChildren : <Icon as={AddIcon} />}
              </Button>
            )}
          </VStack>
        </Flex>
      ))}
    </Flex>
  );
};

export default ArrayField;
