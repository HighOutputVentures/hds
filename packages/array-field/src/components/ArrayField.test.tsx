import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import '@testing-library/react/dont-cleanup-after-each';
import React from 'react';
import renderer from 'react-test-renderer';
import ArrayField from './ArrayField';

describe('Array Field Component', () => {
  beforeAll(() => {
    render(
      <ArrayField
        defaultValues={{
          input: [
            {
              value: '',
            },
          ],
        }}
        maxValue={5}
        onChange={jest.fn()}
        onRemove={jest.fn()}
        onAppend={jest.fn()}
        onBlur={jest.fn()}
      />
    );
  });

  it('should renders input field form container', () => {
    const formControl = screen.queryByTestId('hds.form.control');
    expect(formControl).toBeDefined();
  });

  it('should renders input field input group', () => {
    const inputGroup = screen.queryByTestId('hds.input.group');
    expect(inputGroup).toBeDefined();
  });

  it('should renders input field input', () => {
    const input = screen.queryByTestId('hds.input.field');
    expect(input).toBeDefined();
  });

  describe('Snapshot', () => {
    it('Should match snapshot', () => {
      const component = renderer.create(
        <ArrayField
          defaultValues={{
            input: [{ value: '' }],
          }}
          maxValue={3}
          onChange={function () {}}
          onRemove={function () {}}
          onAppend={function () {}}
          onBlur={function () {}}
        />
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      component.unmount();
    });
  });
});
