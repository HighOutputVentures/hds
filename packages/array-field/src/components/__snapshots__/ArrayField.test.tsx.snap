// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array Field Component Snapshot Should match snapshot 1`] = `
<div
  className="css-3a3hw0"
>
  <div
    className="css-147dy5z"
    onBlur={[Function]}
    onChange={[Function]}
  >
    <div
      className="chakra-form-control css-0"
      data-testid="hds.form-group.group"
      role="group"
    >
      <div
        className="chakra-input__group css-4302v8"
      >
        <input
          className="chakra-input css-ay2syu"
          data-testid="hds.array-field.input"
          disabled={false}
          id="input"
          name="input.0.value"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
        />
      </div>
    </div>
    <div
      className="chakra-stack css-owjkmg"
    >
      <button
        className="chakra-button css-1xxv165"
        data-testid="hds.array-field.remove.btn"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <svg
          className="chakra-icon chakra-icon css-ikjmp3"
          focusable={false}
          viewBox="0 0 24 24"
        >
          <g
            fill="currentColor"
          >
            <path
              d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
            />
          </g>
        </svg>
      </button>
      <button
        className="chakra-button css-1xxv165"
        data-testid="hds.array-field.add.btn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          className="chakra-icon chakra-icon css-ikjmp3"
          focusable={false}
          viewBox="0 0 24 24"
        >
          <path
            d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
